### ISSUE TITLE
StringIndexOutOfBoundsException

### ISSUE THREAD

**Issue by damercier** [stefan-niedermann/nextcloud-notes#908]
Hello,

I got the following StringIndexOutOfBoundsException when playing with checkboxes in preview mode.

**Steps to reproduce:**
- create a note
- add a list with checkboxes
- go to preview mode
- check and uncheck quickly the same checkbox (latest?)

```
App Version: 2.16.3
App Version Code: 2016003
App Flavor: fdroid

Files App Version Code: 30120090

---

OS Version: 4.1.18-g4ae3101(eng.jslave.20170320.182011)
OS API Level: 24
Device: HWVNS-H
Manufacturer: HUAWEI
Model (and Product): HUAWEI VNS-L31 (VNS-L31)

---

java.lang.StringIndexOutOfBoundsException
	at android.text.SpannableStringBuilder.<init>(SpannableStringBuilder.java:60)
	at android.text.SpannableStringBuilder.subSequence(SpannableStringBuilder.java:1143)
	at com.yydcdut.markdown.utils.TextHelper.formatTodoLine(TextHelper.java:208)
	at com.yydcdut.markdown.syntax.text.TodoSyntax.onTodoClicked(TodoSyntax.java:93)
	at com.yydcdut.markdown.utils.SyntaxUtils$1.onClick(SyntaxUtils.java:266)
	at android.text.method.LinkMovementMethod.onTouchEvent(LinkMovementMethod.java:217)
	at android.widget.TextView.onTouchEvent(TextView.java:8827)
	at android.view.View.dispatchTouchEvent(View.java:10012)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2671)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2358)
	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:447)
	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1871)
	at android.app.Activity.dispatchTouchEvent(Activity.java:3213)
	at androidx.appcompat.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:69)
	at androidx.appcompat.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:69)
	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:409)
	at android.view.View.dispatchPointerEvent(View.java:10232)
	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:4856)
	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4717)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4249)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:4302)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:4268)
	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:4395)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:4276)
	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:4452)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4249)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:4302)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:4268)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:4276)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4249)
	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:6676)
	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:6650)
	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:6611)
	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:6803)
	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:192)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:330)
	at android.os.Looper.loop(Looper.java:138)
	at android.app.ActivityThread.main(ActivityThread.java:6523)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:941)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:831)

```

**Comment by stefan-niedermann** [comment]
Thank you for the report.

Can you please post a minimal sample to reproduce this issue?

**Comment by stefan-niedermann** [comment]
@damercier The root cause is in an library (which we already plan to replace and cannot patch easily).

I can only try to catch the error in a workaround but this is only possible if i can reproduce this issue.

For reproduction i need a minimal sample note where this error appears. Can you please give me this kind of sample note where you can reproduce the error?

Otherwise i can't do anything unfortunately.

**Comment by stefan-niedermann** [comment]
[Why this bug report has been closed.](https://github.com/stefan-niedermann/nextcloud-notes/blob/master/FAQ.md#why-has-my-bug-report-been-closed)

**Comment by damercier** [comment]
Hello @stefan-niedermann, 
Sorry for the delay, here is a minimal sample note where the error appears. You have to click multiple time [quickly?], it's not always reproducible unfortunately…

```
# Test

- [ ] test1
- [ ] test2
- [ ]
``` 

After some tests, it seems that the last empty line is "needed" to get the bug… 

A `try/catch` as workaround can probably already help for the stability of the Notes application.

**Comment by stefan-niedermann** [comment]
> A try/catch as workaround can probably already help

Not that easy, unfortunately. The exception comes from a not maintained 3rd-party library which i do not have direct access to. I will see what i can do though.

**Comment by stefan-niedermann** [comment]
Fixed in `2.19.0`.
